### PR TITLE
Repo search

### PR DESCRIPTION
**Pull request by grosser**
### Description
- Search through Repo URL
/cc @zendesk/samson

### References
 - Jira link: https://zendesk.atlassian.net/browse/BRE-915

### Risks
- Level: Low

### CODE REVIEW

File: app/controllers/projects_controller.rb [hunk 1]
```diff
@@ -112,8 +112,22 @@ def project_params
...
-        Project.search(query)
+      if search = params.dig(:search).presence
+        scope = Project
+        if query = search[:query]
+          scope = scope.search(query)
+        end
+        if url = search[:url]
+          # we store urls as git@github.com:foo/bar.git or https://github.com/foo/bar.git
+          # but support search with either and also https://github.com/foo/bar
+          _, path = url.split(/\.[a-z]+[\/:]/, 2)
```
Comment (ragurney) [inline app/controllers/projects_controller.rb]: I kind of prefer `path = url.split(/\.[a-z]+[\/:]/, 2)[1]` 🤷‍♂️

File: test/controllers/projects_controller_test.rb [hunk 2]
```diff
@@ -34,10 +34,31 @@ def fields_disabled?
...
+          get :index, params: {search: {query: "a"}}
+          assigns(:projects).map(&:name).must_equal ["a"]
+        end
+
+        it "url in params" do
+          get :index, params: {search: {url: "git@github.com/foo/bar.git"}}
+          assigns(:projects).map(&:name).must_equal ["a", "b"]
+        end
+
+        it "both query and url in params" do
```
Comment (grosser) [inline test/controllers/projects_controller_test.rb]: `can combine query and url`

File: test/controllers/projects_controller_test.rb [hunk 3]
```diff
@@ -34,10 +34,31 @@ def fields_disabled?
...
-      it "can search" do
-        Project.create!(name: "a", repository_url: "a")
-        get :index, params: {search: {query: "o"}}
-        assigns(:projects).map(&:name).must_equal ["Foo"]
+      describe "can search" do
+        before do
+          Project.create!(name: "a", repository_url: "git@github.com/foo/bar.git")
+          Project.create!(name: "b", repository_url: "https://github.com/foo/bar.git")
+        end
+        it "query in params" do
```
Comment (grosser) [inline test/controllers/projects_controller_test.rb]: build real sentences: `it can search via query`

Comment (grosser) [reply test/controllers/projects_controller_test.rb]: also newline above

File: test/controllers/projects_controller_test.rb [hunk 4]
```diff
@@ -34,10 +34,31 @@ def fields_disabled?
...
+        it "url in params" do
+          get :index, params: {search: {url: "git@github.com/foo/bar.git"}}
+          assigns(:projects).map(&:name).must_equal ["a", "b"]
+        end
+
+        it "both query and url in params" do
+          get :index, params: {search: {query: "b", url: "https://github.com/foo/bar.git"}}
+          assigns(:projects).map(&:name).must_equal ["b"]
+        end
+
```
Comment (grosser) [inline test/controllers/projects_controller_test.rb]: negative test for "it does not find when url does not match"

File: test/controllers/projects_controller_test.rb [hunk 5]
```diff
@@ -34,10 +34,31 @@ def fields_disabled?
...
       end
 
-      it "can search" do
-        Project.create!(name: "a", repository_url: "a")
-        get :index, params: {search: {query: "o"}}
-        assigns(:projects).map(&:name).must_equal ["Foo"]
+      describe "can search" do
+        before do
+          Project.create!(name: "a", repository_url: "git@github.com/foo/bar.git")
+          Project.create!(name: "b", repository_url: "https://github.com/foo/bar.git")
```
Comment (grosser) [inline test/controllers/projects_controller_test.rb]: we already have 2 fixture projects ... can we reuse them instead ?

**Comment by grosser** [comment]
update to parse the url and explicitly generate all supported urls ... looks good ?
(possibly breaks when users figure out a new format, but it's less magical and always supports exact matches)

**Pull request merged**